### PR TITLE
netplan: ensure ethernets is first key when rendering v2 passthrough

### PR DESCRIPTION
**Pull request by raharper**
## Proposed Commit Message

```
netplan: ensure ethernets is first key when rendering v2 passthrough

Render netplan content as passthrough but ensure that 'ethernets' is the first
key in the final netplan content. netplan parser will fail if bonds, bridges,
and vlans reference links which have not be parsed

LP: #2040295
```

## Additional Context

`netplan generate` and/or `netplan apply` have had issues with referencing
interfaces that have yet to be parsed ( at least in focal and jammy); it may
be the case that newer netplan releases have resolved this issue.

## Test Steps

```
$ lxc launch ubuntu-daily:jammy j2
$ lxc exec j2 /bin/bash
cat >test.yaml <<EOF
network:
  version: 2
  ethernets:
    eth1-1:
      match:
        name: "eth1-1"
      dhcp4: true
    eth1-2:
      match:
        name: "eth1-2"
      dhcp4: true
  bridges:
    oobmgmt:
      addresses: [10.24.232.207/23]
      interfaces: []
EOF
$ cloud-init devel net-convert --network-data test.yaml -k yaml -d nplantest -D ubuntu -O netplan --debug
$ cat nplantest/etc/netplan/50-cloud-init.yaml
# This file is generated from information provided by the datasource.  Changes
# to it will not persist across an instance reboot.  To disable cloud-init's
# network configuration capabilities, write a file
# /etc/cloud/cloud.cfg.d/99-disable-network-config.cfg with the following:
# network: {config: disabled}
network:
    bridges:
        oobmgmt:
            addresses:
            - 10.24.232.207/23
            interfaces: []
    ethernets:
        eth1-1:
            dhcp4: true
            match:
                name: eth1-1
        eth1-2:
            dhcp4: true
            match:
                name: eth1-2
    version: 2

```

## Checklist
- [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [x] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

### CODE REVIEW

**Review by TheRealFalcon — COMMENTED** [review]
I left an inline comment if we continue with this solution, but doing our own rendering makes me a little nervous. Is this a problem you have seen recently in netplan? I thought it was solved [a while ago](https://github.com/canonical/netplan/commit/e41215b24f43f5a67ba9f38bd6373a0f75405335).

If this is necessary, what do you think about doing something like this instead?
```python
        new_config = {}
        if "version" in config:
            new_config["version"] = config.pop("version")
        if "ethernets" in config:
            new_config["ethernets"] = config.pop("ethernets")
        new_config.update(config)

        return safeyaml.dumps(
            {"network": new_config},
            explicit_start=False,
            explicit_end=False,
            noalias=True,
            sort_keys=False,
        )
```

Then we can add an option for `sort_keys=False` on the dump call. Since insertion order is guaranteed on Python now, it allows to sidestep doing our own rendering.

**Comment by raharper** [comment]
> I left an inline comment if we continue with this solution, but doing our
> own rendering makes me a little nervous. Is this a problem you have seen
> recently in netplan? I thought it was solved [a while
> ago](https://github.com/canonical/netplan/commit/e41215b24f43f5a67ba9f38bd6373a0f75405335)
> (pre-focal).

We definitely observed in focal.  I'll follow up if current focal and jammy
can reproduce the netplan issue with an out-of-order config.

>
> If this is necessary, what do you think about doing something like this instead?
>
> ```python
>         new_config = {}
>         if "version" in config:
>             new_config["version"] = config.pop("version")
>         if "ethernets" in config:
>             new_config["ethernets"] = config.pop("ethernets")
>         new_config.update(config)
>
>         return safeyaml.dumps(
>             {"network": new_config},
>             explicit_start=False,
>             explicit_end=False,
>             noalias=True,
>             sort_keys=False,
>         )
> ```
>
> Then we can add an option for `sort_keys=False` on the dump call. Since
> insertion order is guaranteed on Python now, it allows to sidestep doing our
> own rendering.

This seems reasonable to merge into tip here; the remaining downstream
question is what is more easily backportable to say python 3.8 in focal; does
that version also ensure insertion order is retained?

File: cloudinit/net/netplan.py [hunk 1]
```diff
@@ -352,16 +366,27 @@ def _net_setup_link(self, run=False):
...
 
+    def _render_passthrough(self, config):
+        """Render netplan content as passthrough but ensure that
+        'ethernets' is the first key in the final netplan content. netplan
+        parser will fail if bond, bridges, vlans reference links which have
+        not be parsed.
+        """
+        content = []
+        content.append("network:\n    version: 2\n")
+        to_render = ["ethernets"] + [
```
[the remark this answers is not in the document]
Comment (raharper) [reply cloudinit/net/netplan.py:377]: OK

**Comment by raharper** [comment]
> > I left an inline comment if we continue with this solution, but doing our
> > own rendering makes me a little nervous. Is this a problem you have seen
> > recently in netplan? I thought it was solved [a while
> > ago](https://github.com/canonical/netplan/commit/e41215b24f43f5a67ba9f38bd6373a0f75405335)
> > (pre-focal).
>
> We definitely observed in focal. I'll follow up if current focal and jammy can reproduce the netplan issue with an out-of-order config.

```
root@f2test:/etc/netplan# apt-cache policy cloud-init
cloud-init:
  Installed: 23.3.1-0ubuntu1~20.04.1
  Candidate: 23.3.1-0ubuntu1~20.04.1
  Version table:
 *** 23.3.1-0ubuntu1~20.04.1 500
        500 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
        100 /var/lib/dpkg/status
     23.1.2-0ubuntu0~20.04.2 500
        500 http://security.ubuntu.com/ubuntu focal-security/main amd64 Packages
     20.1-10-g71af48df-0ubuntu5 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
root@f2test:/etc/netplan# apt-cache policy netplan.io
netplan.io:
  Installed: 0.104-0ubuntu2~20.04.2
  Candidate: 0.104-0ubuntu2~20.04.2
  Version table:
 *** 0.104-0ubuntu2~20.04.2 500
        500 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages
        100 /var/lib/dpkg/status
     0.99-0ubuntu1 500
        500 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages
root@f2test:/etc/netplan# netplan info
netplan.io:
  website: "https://netplan.io/"
  features:
  - generate-just-in-time
  - generated-supplicant
  - dbus-config
  - default-routes
  - auth-phase2
  - dhcp-use-domains
  - ipv6-mtu
  - modems
  - sriov
  - openvswitch
  - activation-mode
  - eswitch-mode
```

This version of netplan does not complain and appears to bring up all of the
various links.

```
network:
    version: 2
    bonds:
        bond0:
            link-local: []
            mtu: 1500
            dhcp4: false
            dhcp6: false
            accept-ra: false
            parameters:
                mode: active-backup
                mii-monitor-interval: 60
            interfaces:
                - eth0
        bond1:
            link-local: []
            mtu: 1500
            dhcp4: false
            dhcp6: false
            accept-ra: false
            parameters:
                mode: active-backup
                mii-monitor-interval: 60
            interfaces:
                - eth1
    bridges:
        lxcbr0:
            addresses:
                - 169.254.254.254/32
            link-local: []
            mtu: 1500
            dhcp4: false
            dhcp6: false
            accept-ra: false
            parameters:
                stp: false
                forward-delay: 0
            interfaces:
                - bond1
        oobmgmt:
            addresses:
                - 10.24.232.207/23
            link-local: []
            mtu: 1500
            dhcp4: false
            dhcp6: false
            accept-ra: false
            parameters:
                stp: false
                forward-delay: 0
            interfaces:
                - bond0
    vlans:
        bond0.1:
            id: 1
            link: bond0
            addresses:
                - 10.0.0.1/32
            # ARPCHECK=yes implemented elsewhere
            link-local: []
            mtu: 1496
            dhcp4: false
            dhcp6: false
            accept-ra: false

    ethernets:
        eth0:
            dhcp4: false
        eth1:
            dhcp4: false
```

configured interfaces

```
root@f2test:/etc/netplan# ip link show
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: lxcbr0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 46:b2:b5:5c:d8:f0 brd ff:ff:ff:ff:ff:ff
3: oobmgmt: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff
4: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master oobmgmt state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff
5: bond1: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master lxcbr0 state UP mode DEFAULT group default qlen 1000
    link/ether 46:b2:b5:5c:d8:f0 brd ff:ff:ff:ff:ff:ff
6: bond0.1@bond0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1496 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff
58: eth0@if59: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc noqueue master bond0 state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff link-netnsid 0
60: eth1@if61: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc noqueue master bond1 state UP mode DEFAULT group default qlen 1000
    link/ether 46:b2:b5:5c:d8:f0 brd ff:ff:ff:ff:ff:ff link-netnsid 0

root@f2test:/etc/netplan# ip link show type bridge_slave
4: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master oobmgmt state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff
5: bond1: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue master lxcbr0 state UP mode DEFAULT group default qlen 1000
    link/ether 46:b2:b5:5c:d8:f0 brd ff:ff:ff:ff:ff:ff

root@f2test:/etc/netplan# ip link show type bond_slave
58: eth0@if59: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc noqueue master bond0 state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff link-netnsid 0
60: eth1@if61: <BROADCAST,MULTICAST,SLAVE,UP,LOWER_UP> mtu 1500 qdisc noqueue master bond1 state UP mode DEFAULT group default qlen 1000
    link/ether 46:b2:b5:5c:d8:f0 brd ff:ff:ff:ff:ff:ff link-netnsid 0

root@f2test:/etc/netplan# ip link show type vlan
6: bond0.1@bond0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1496 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether 4e:73:6a:9a:53:87 brd ff:ff:ff:ff:ff:ff
```

I suspect then the issue is resolved with newer netplan (yay).  I think we can close this PR.

Would you consider a PR to do like you suggested (sort_keys=false) in the pass-through case?  And possibly not emit the cloud-init header (though that's not a big deal).

**Comment by TheRealFalcon** [comment]
> the remaining downstream question is what is more easily backportable to say python 3.8 in focal; does that version also ensure insertion order is retained?

Yes, it goes back to 3.6 as an implementation detail and 3.7 as part of the language spec.

> Would you consider a PR to do like you suggested (sort_keys=false) in the pass-through case? And possibly not emit the cloud-init header (though that's not a big deal).

What would be the reason? I don't want to add code that winds up being superfluous, but as long as there's a reason that doesn't negatively affect others, it should be accepted.

**Comment by raharper** [comment]
> > the remaining downstream question is what is more easily backportable to say python 3.8 in focal; does that version also ensure insertion order is retained?
>
> Yes, it goes back to 3.6 as an implementation detail and 3.7 as part of the language spec.

Excellent.

>
> > Would you consider a PR to do like you suggested (sort_keys=false) in the pass-through case? And possibly not emit the cloud-init header (though that's not a big deal).
>
> What would be the reason? I don't want to add code that winds up being superfluous, but as long as there's a reason that doesn't negatively affect others, it should be accepted.

The name "passthrough" suggests that cloud-init doesn't modify the input.
That's mostly true except we:

1) re-order the yaml keys
2) lose any comments
3) write a header at the top

**Comment by TheRealFalcon** [comment]
Ah right. Pure passthrough might be difficult due to how early we fetch the config, but yes, I agree that it is a good idea.

**Comment by TheRealFalcon** [comment]
@raharper , I am closing this as you mentioned the other work happening as a separate PR. If there is still more to do here, feel free to re-open.

**Pull request closed without merging**